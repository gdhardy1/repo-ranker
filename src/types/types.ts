import { Dispatch, SetStateAction } from "react";
import { LazyQueryResult } from "@apollo/client/react/types/types";
export interface OrganizationData {
  organization: Organization;
}

export interface Organization {
  id: string;
  login: string;
  name: string;
  repositories: {
    edges: {
      node: RepositoryNode;
    }[];
    totalCount: number;
    pageInfo: PageInfo;
  };
}

export interface RepositoryNode {
  name: string;
  defaultBranchRef: { name: string };
  id: string;
  stargazers: { totalCount: number };
}
export interface RepositoryData {
  repository: Repository;
}

export interface Repository {
  name: string;
  object: CommitHistory; // git objects filtered by Git rev-parse expression
}

export interface CommitHistory {
  history: {
    totalCount: number;
    edges: {
      node: CommitNode;
    }[];
    pageInfo: PageInfo;
  };
}

export interface CommitNode {
  id: string;
  messageHeadline: string;
  url: string;
}

export interface PageInfo {
  startCursor: string;
  endCursor: string;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
}

export class NullOrganizationData implements OrganizationData {
  organization;

  constructor() {
    this.organization = {
      id: "",
      login: "",
      name: "",
      repositories: {
        edges: [] as { node: RepositoryNode }[],
        totalCount: 0,
        pageInfo: {
          startCursor: "",
          endCursor: "",
          hasNextPage: false,
          hasPreviousPage: false,
        },
      },
    };
  }
}

export class NullRepositoryData implements RepositoryData {
  repository;
  constructor() {
    this.repository = {
      name: "",
      object: {
        history: {
          totalCount: 0,
          edges: [
            {
              node: {
                id: "",
                messageHeadline: "",
                url: "",
              },
            },
          ],
          pageInfo: {
            startCursor: "",
            endCursor: "",
            hasNextPage: false,
            hasPreviousPage: false,
          },
        },
      },
    };
  }
}
export interface OrganizationDataVariables {
  login: string;
  cursor?: string | undefined;
}
export interface RepositoryDataVariables {
  login: string;
  repoName: string;
  branch: string;
  cursor?: string;
}

export type LazyGetter<S, T> = (
  variables: T
) => Promise<LazyQueryResult<S, T> | undefined>;

export type LazyGetterTuple<S, T> = [LazyGetter<S, T>, LazyQueryResult<S, T>];

export type AppContextType = {
  organizationData: OrganizationData;
  setOrganizationData: Dispatch<SetStateAction<OrganizationData>>;
  repositoryData: RepositoryData;
  setRepositoryData: Dispatch<SetStateAction<RepositoryData>>;
  nextCursor: string | undefined;
  setNextCursor: Dispatch<SetStateAction<string | undefined>>;
  notification: Notification;
  setNotification: Dispatch<SetStateAction<Notification>>;
};

export interface Notification {
  message: string;
  title: string;
  show: boolean;
}
