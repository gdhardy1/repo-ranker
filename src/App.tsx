import React, { useState, createContext } from "react";
import "./App.css";
import {
  AppContextType,
  OrganizationData,
  RepositoryData,
  NullOrganizationData,
  NullRepositoryData,
} from "./types/types";
import Search from "./components/Search";
import { Outlet } from "react-router-dom";

export const AppContext = createContext({} as AppContextType);

function App() {
  const [organizationData, setOrganizationData] = useState<OrganizationData>(
    new NullOrganizationData()
  );
  const [repositoryData, setRepositoryData] = useState<RepositoryData>(
    new NullRepositoryData()
  );
  const [startCursor, setStartCursor] = useState("");
  const [nextCursor, setNextCursor] = useState<string | undefined>(undefined);
  const [totalRepoCount, setTotalRepoCount] = useState(0);
  const [login, setLogin] = useState<string>("");
  const context = {
    organizationData,
    setOrganizationData,
    repositoryData,
    setRepositoryData,
    nextCursor,
    setNextCursor,
  };

  return (
    <div className="App">
      <AppContext.Provider value={context}>
        <Search />
        <Outlet />
      </AppContext.Provider>
    </div>
  );
}

export default App;
