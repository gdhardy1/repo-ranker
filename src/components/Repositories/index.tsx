import React, { useContext } from "react";
import { useNavigate, useParams } from "react-router-dom";
import { AppContext } from "../../App";
import { StarIcon } from "@heroicons/react/20/solid";

export default function Repositories() {
  const { organizationData } = useContext(AppContext);
  const navigate = useNavigate();
  const params = useParams();

  const handleViewCommits = async (variables: any) => {
    navigate(
      `/${params.organizationLogin}/${variables.repoName}/${variables.branch}`
    );
  };
  return (
    <>
      {organizationData.organization.repositories.edges.map((edge: any) => {
        const { node: repository } = edge;
        return (
          <li
            className="repository-card"
            key={repository.id}
            onClick={(event) => {
              event.preventDefault();
              handleViewCommits({
                login: organizationData.organization.login,
                repoName: repository.name,
                branch: repository.defaultBranchRef.name,
              });
            }}
          >
            <div className="block hover:bg-gray-50">
              <div className="px-4 py-4 sm:px-6">
                <div className="flex items-center justify-between">
                  <p className="truncate text-sm font-medium text-indigo-600">
                    {repository.name}
                  </p>
                  <div className="ml-2 flex flex-shrink-0">
                    <p className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800">
                      {repository.stargazers.totalCount}
                      <StarIcon
                        className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                        aria-hidden="true"
                      />
                    </p>
                  </div>
                </div>
              </div>
            </div>
          </li>
        );
      })}
    </>
  );
}
