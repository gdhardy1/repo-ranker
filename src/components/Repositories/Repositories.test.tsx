import { render, screen } from "@testing-library/react";

import "@testing-library/jest-dom";
import Repositories from ".";
import { AppContext } from "../../App";
import * as mocks from "../../util/test/mocks";
import { AppContextType } from "../../types/types";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

describe("<Repositories />", () => {
  test("Renders correctly", async () => {
    const mockContext: AppContextType = {
      organizationData: mocks.organizationData,
      repositoryData: mocks.repositoryData,
      nextCursor: "",
      notification: { title: "", message: "", show: false },
      setNextCursor: jest.fn(),
      setOrganizationData: jest.fn(),
      setRepositoryData: jest.fn(),
      setNotification: jest.fn(),
    };

    const tree = renderer
      .create(
        <BrowserRouter>
          <AppContext.Provider value={mockContext}>
            <Repositories />
          </AppContext.Provider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
