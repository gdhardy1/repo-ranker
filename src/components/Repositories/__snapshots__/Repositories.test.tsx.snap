// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repositories /> Renders correctly 1`] = `
Array [
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              53590
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-starter-blog
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              3343
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-starter-default
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              1356
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-starter-hello-world
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              630
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            store.gatsbyjs.org
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              594
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-source-wordpress-experimental
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              389
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-docker
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              352
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-starter-shopify
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              262
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            desktop
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              222
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
  <li
    className="repository-card"
    onClick={[Function]}
  >
    <div
      className="block hover:bg-gray-50"
    >
      <div
        className="px-4 py-4 sm:px-6"
      >
        <div
          className="flex items-center justify-between"
        >
          <p
            className="truncate text-sm font-medium text-indigo-600"
          >
            gatsby-pt-BR
          </p>
          <div
            className="ml-2 flex flex-shrink-0"
          >
            <p
              className="inline-flex rounded-full bg-green-100 px-2 text-xs font-semibold leading-5 text-green-800"
            >
              153
              <svg
                aria-hidden="true"
                className="h-4 w-4 ml-2 my-auto flex-shrink-0"
                fill="currentColor"
                viewBox="0 0 20 20"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clipRule="evenodd"
                  d="M10.868 2.884c-.321-.772-1.415-.772-1.736 0l-1.83 4.401-4.753.381c-.833.067-1.171 1.107-.536 1.651l3.62 3.102-1.106 4.637c-.194.813.691 1.456 1.405 1.02L10 15.591l4.069 2.485c.713.436 1.598-.207 1.404-1.02l-1.106-4.637 3.62-3.102c.635-.544.297-1.584-.536-1.65l-4.752-.382-1.831-4.401z"
                  fillRule="evenodd"
                />
              </svg>
            </p>
          </div>
        </div>
      </div>
    </div>
  </li>,
]
`;
